Command-line parsing: More helpful messages on failed parsing

It's reasonably common for users to get confused by the `[ERROR] expected exactly # arguments (# provided)` error.

A relatively common mistake is to omit the preceding dash from options, so that the option (and any subsequent parameters to that option) get interpreted as arguments. When this occurs, it should be possible to search for the presence of command-line options where the option name matches the text provided in one of the arguments, and therefore give the user a hint that they may have forgotten the '-'.

Another possibility to help here would be to print out the command usage _with arguments only shown_; and then print out what the user has provided as arguments ie. _with the provided options & parameters stripped_; that might help users isolate where they've mucked up their command string...?

--- a/lib/app.cpp
+++ b/lib/app.cpp
@@ -907,12 +907,35 @@ namespace MR
       }
 
       if (num_optional_arguments && num_args_required > argument.size())
-        throw Exception ("expected at least " + str (num_args_required)
+        throw Exception ("Expected at least " + str (num_args_required)
             + " arguments (" + str (argument.size()) + " supplied)");
 
-      if (num_optional_arguments == 0 && num_args_required != argument.size())
-        throw Exception ("expected exactly " + str (num_args_required)
+      if (num_optional_arguments == 0 && num_args_required != argument.size()) {
+        Exception e ("Expected exactly " + str (num_args_required)
             + " arguments (" + str (argument.size()) + " supplied)");
+        std::string s = "Usage: " + NAME;
+        for (const auto& a : ARGUMENTS)
+          s += " " + std::string(a.id);
+        e.push_back (s);
+        s = "Yours: " + NAME;
+        for (const auto& a : argument)
+          s += " " + std::string(a);
+        e.push_back (s);
+        if (argument.size() > num_args_required) {
+          std::vector<std::string> potential_options;
+          for (const auto& a : argument) {
+            for (const auto& og : OPTIONS) {
+              for (const auto& o : og) {
+                if (std::string(a) == std::string(o.id))
+                  potential_options.push_back ("'-" + a + "'");
+              }
+            }
+          }
+          if (potential_options.size())
+            e.push_back ("(Did you mean " + join(potential_options, " or ") + "?)");
+        }
+        throw e;
+      }
 
       size_t num_extra_arguments = argument.size() - num_args_required;
       size_t num_arg_per_multi = num_optional_arguments ? num_extra_arguments / num_optional_arguments : 0;

--- a/lib/exception.h
+++ b/lib/exception.h
@@ -98,6 +98,9 @@ namespace MR
       const std::string& operator[] (size_t n) const {
         return description[n];
       }
+      void push_back (const std::string& s) {
+        description.push_back (s);
+      }
 
       static void (*display_func) (const Exception& E, int log_level);
 

